pennylane/transforms/decompositions/clifford_t/clifford_t_transform.py

```diff
@@ -0,0 +1,479 @@
+@transform
+def clifford_t_decomposition(
+    tape: QuantumTape,
+    epsilon=1e-4,
+    max_expansion=6,
+    method="sk",
+    **method_kwargs,
+) -> (Sequence[QuantumTape], Callable):
+    r"""Unrolls a circuit into the Clifford+T basis using the optimal ancilla-free approximation of :class:`~.RZ` operations.
+
+    This method first decomposes the gate operations to a basis comprised of Clifford, :class:`~.T`, :class:`~.RZ` and
+    :class:`~.GlobalPhase` operations (and their adjoints). The Clifford gates include the following PennyLane operations:
+
+    - Single qubit gates - :class:`~.Identity`, :class:`~.PauliX`, :class:`~.PauliY`, :class:`~.PauliZ`,
+      :class:`~.SX`, :class:`~.S`, and :class:`~.Hadamard`.
+    - Two qubit gates - :class:`~.CNOT`, :class:`~.CY`, :class:`~.CZ`, :class:`~.SWAP`, and :class:`~.ISWAP`.
+
+    Then the leftover single qubit :class:`~.RZ` operations are approximated in the Clifford+T basis with
+    :math:`\epsilon > 0` error using the Solovay-Kitaev algorithm described in
+    `Dawson and Nielsen (2005) <https://arxiv.org/abs/quant-ph/0505030>`_.
+
+    Args:
+        tape (QuantumTape): A quantum tape
+        epsilon (float): The maximum permissible error. Defaults to ``0.0001``.
+        max_expansion (int): The depth to be used for tape expansion before manual decomposition to Clifford+T basis is applied.
+        method (str): Method to be used for Clifford+T decomposition. Default value is ``"sk"`` for Solovay-Kitaev
+        **method_kwargs: Keyword argument to pass options for the ``method`` used for decompositions.
+
+    Keyword Args:
+        Options (*):
+
+            * **max_depth** (int), **basis_set** (list[str]), **basis_length** (int):
+              arguments for the ``"sk"`` method, i.e., for performing Solovay-Kitaev decomposition using :func:`~.sk_decomposition`
+
+    Returns:
+        pennylane.QNode or qfunc or tuple[List[.QuantumTape], function]: If a QNode is passed,
+        it returns a QNode with the transform added to its transform program. If a tape is passed,
+        returns a tuple containing a list of quantum tapes to be evaluated, and a function to be
+        applied to these tape executions.
+
+    Raises:
+        ValueError: If any gate operation does not have a rule for its decomposition
+        NotImplementedError: If chosen decomposition is not supported
+
+    .. seealso:: :func:`~.sk_decomposition` for Solovay-Kitaev decomposition.
+
+    **Example**
+
+    .. code-block:: python3
+
+        @qml.qnode(qml.device("default.qubit"))
+        def circuit(x, y):
+            qml.RX(x, 0)
+            qml.CNOT([0, 1])
+            qml.RY(y, 0)
+            return qml.expval(qml.PauliZ(0))
+
+        x, y = 1.1, 2.2
+        decomposed_circuit = qml.transforms.clifford_t_decomposition(circuit)
+        result = circuit(x, y)
+        approx = decomposed_circuit(x, y)
```

Comment:
Dare I ask for a drawing for this? Would it be very deep :laughing:

Reply:
Thanks! An alternative is to query the resource requirements, but I'm also fine with keeping it simple here.

Reply:
tbh I wouldn't want to draw too much attention to it, especially in high-level docs, because the results would be... disappointing. but perhaps we could add it somewhere else later?